ehrql/serializer.py

```diff
@@ -0,0 +1,265 @@
+        pathlib.Path,
+        ColumnSpec,
+        DummyDataConfig,
+        Measure,
+        *get_all_subclasses(Node),
+        *get_all_subclasses(BaseCode),
+        *get_all_subclasses(BaseConstraint),
+        *get_all_subclasses(BaseDatasetReader),
+    ]
+}
+
+TYPE_REGISTRY_INVERSE = {type_: name for name, type_ in TYPE_REGISTRY.items()}
+
+# Ensure no name clashes
+assert len(TYPE_REGISTRY) == len(TYPE_REGISTRY_INVERSE)
+
+
+def type_name(obj):
+    return TYPE_REGISTRY_INVERSE[type(obj)]
+
+
+def serialize(value):
+    return json.dumps(Marshaller.to_dict(value), indent=2)
+
+
+def deserialize(data):
+    return Unmarshaller.from_dict(json.loads(data))
+
+
+class Marshaller:
+    """
+    Convert arbitrary values to nested structures of JSON-supported values ready to be
+    serialized
+    """
+
+    @classmethod
+    def to_dict(cls, value):
+        marshaller = cls()
+        return {
+            "value": marshaller.marshal(value),
+            **marshaller.get_references(),
+        }
+
+    def __init__(self):
+        self.references = {}
+        # Sequence of unique IDs (strings not ints so they can be used as JSON keys)
+        id_sequence = (str(i) for i in range(1, 2**32))  # pragma: no branch
+        self.get_next_reference_id = id_sequence.__next__
+
+    def get_references(self):
+        return dict(self.references.values())
+
+    @functools.singledispatchmethod
+    def marshal(self, obj):
+        assert False, f"Unsupported type '{type(obj)}': {obj!r}"
+
+    @marshal.register(type(None))
+    @marshal.register(bool)
+    @marshal.register(str)
+    @marshal.register(int)
+    @marshal.register(float)
+    def marshal_primitive(self, obj):
+        return obj
+
+    @marshal.register(datetime.date)
+    def marshal_date(self, obj):
+        return {type_name(obj): obj.isoformat()}
+
+    @marshal.register(list)
+    def marshal_list(self, obj):
+        return [self.marshal(value) for value in obj]
+
+    @marshal.register(tuple)
+    @marshal.register(frozenset)
+    def marshal_iterable(self, obj):
+        return {type_name(obj): [self.marshal(value) for value in obj]}
+
+    @marshal.register(dict)
+    def marshal_dict(self, obj):
+        return {
+            type_name(obj): [
+                [self.marshal(key), self.marshal(value)] for key, value in obj.items()
+            ],
+        }
+
+    @marshal.register(type)
+    def marshal_type(self, obj):
+        return {"type": TYPE_REGISTRY_INVERSE[obj]}
+
+    @marshal.register(pathlib.Path)
+    def marshal_path(self, obj):
+        # Paths are unusual in that we always want to encode it as the platform-agnostic
+        # alias `Path`, not as the specific subclass (e.g. `PosixPath`) it happens to be
+        return {"Path": self.marshal(str(obj))}
+
+    @marshal.register(Value)
+    @marshal.register(Position)
+    @marshal.register(BaseCode)
+    def marshal_value(self, obj):
+        return {type_name(obj): self.marshal(obj.value)}
```

Comment:
I've added some comments here:
```py
    # We need special handling for enums like `Position` as the value attribute doesn't
    # appear inside `__dict__` as it does for other objects
    @marshal.register(Position)
    # `BaseCode` objects _can_ be marshalled like other objects, but we'd rather not
    # leak the implementation detail that their one attribute happens to be called
    # "value" so we marshal them as e.g.
    #
    #   {"SNOMEDCTCode": "123456"}
    #
    # rather than
    #
    #   {"SNOMEDCTCode": {"value": "123456"}}
    #
    @marshal.register(BaseCode)
    # Ditto for `Value` objects
    @marshal.register(Value)
```

Reply:
Fair enough. I'm skeptical of the need to avoid leaking the name of the field just because there happens to be only one of them, but not sufficiently skeptical to try to argue you out of it.